aten/src/ATen/native/Sorting.cpp

```diff
@@ -153,34 +209,34 @@ std::tuple<Tensor&, Tensor&> kthvalue_out_cpu(
     indices.zero_();
     return std::forward_as_tuple(values, indices);
   }
-  auto tmp_values = self.clone();
-  auto tmp_indices = at::empty(self.sizes(), self.options().dtype(kLong));
   AT_DISPATCH_ALL_TYPES(self.scalar_type(), "kthvalue_cpu", [&] {
-    dim_apply(
-        {tmp_values, tmp_indices, values, indices},
-        dim,
-        [&](int64_t i, TensorList tl) {
-          auto tmp_values = tl[0].accessor<scalar_t, 1>();
-          auto tmp_indices = tl[1].accessor<int64_t, 1>();
-          scalar_t* mode_value = tl[2].data<scalar_t>();
-          int64_t* mode_index = tl[3].data<int64_t>();
-          for (int64_t j = 0; j < tmp_indices.size(0); j++) {
-            tmp_indices[j] = j;
-          }
-          // we want NaN to be sorted as top for numpy compatibility
-          quick_select_template(
-              tmp_values,
-              k - 1,
-              [](scalar_t x, scalar_t y) -> bool {
-                return ((_isnan<scalar_t>(x) && !_isnan<scalar_t>(y)) || (x > y));
-              },
-              [&](int64_t i, int64_t j) {
-                std::swap(tmp_values[i], tmp_values[j]);
-                std::swap(tmp_indices[i], tmp_indices[j]);
-              });
-          *mode_value = tmp_values[k - 1];
-          *mode_index = tmp_indices[k - 1];
-        });
+    bool can_use_small_index = self.size(dim) <= std::numeric_limits<int_same_size_t<scalar_t>>::max();
```

Comment:
I'm a little confused by the small index logic here (in particular, I don't see why index size should have anything to do with the size of the scalar type; that seems really weird to me!) An explanatory note would help here.